### PR TITLE
Estimate the total scroll height of the table

### LINKED ISSUE #54: Estimate table height
Currently, the `Table` view renders only the first two screens’ worth of data, rendering more as you scroll. However, this can cause problems with inertial scrolling as well as making the scrollbar’s size misleading. One way to fix this would be to add a spacer `<div>` below the table with a height of `(average table row height) * ((number of rows) - (number of rendered rows))`.
Assuming table rows are of similar sizes (which should be correct in most cases) it will result in an accurate estimation of the overall scroll height, and it will only get more accurate as the user scrolls down. A downside to this approach is that if the user scrolls very quickly they may see a flash of white but since you render rows synchronously and fairly quickly this should not be that big fo a problem.

### PR DESCRIPTION
**Pull request by j-f1**
Fixes #54.

This could also be extended to detach or delete rows that are scrolled out of view for full virtualization (although that might be best left to a real virtualization library) since scrolling to the bottom of large lists still freezes the browser.

### CODE REVIEW

File: src/table.js [hunk 1]
```diff
@@ -164,8 +173,9 @@ export function Table(
   }
 
   root.onscroll = () => {
-    if (root.scrollHeight - root.scrollTop < 400 && n < data.length) {
+    while (root.scrollHeight - root.scrollTop - spacer.clientHeight < 400 && n < data.length) {
```
Comment (j-f1) [inline src/table.js:176]: If you scroll really fast, you’ll end up below the newly added rows after one pass, so we keep looping until in-view rows are rendered.

File: src/table.js [hunk 2]
```diff
@@ -194,6 +204,8 @@ export function Table(
 
   revalue();
 
+  setTimeout(updateSpacer, 25);
```
Comment (j-f1) [inline src/table.js:207]: This gives time for Observable to render the cell — I’m not sure if there’s a better or more supported way to defer this until the element is actually rendered. The `clientHeight` values are all zero until the table is placed into the DOM tree.

**Comment by mbostock** [comment]
I think a few improvements are possible here.

1. Use tbody.style.paddingBottom instead of a spacer div.

2. Don’t use clientHeight, and instead compute the height directly from the total number of rows minus the number of displayed rows. We know that every row is exactly 24px tall.

3. Don’t need to use length for input that we control. That’s intended for handling user input and here I would use <code>\`${value}px\`</code> explicitly.

**Comment by j-f1** [comment]
It’s currently 1am here so I will read through my changes again in the morning. Right now I keep two blocks of rows, and shift them around as the user scrolls. This helps preserve things like selection and focus properly, but I’m not sure how it performs. It might be faster to simply re-insert the data into all the rendered rows when we need to swap them.

**Comment by j-f1** [comment]
Updated to reflect the latest changes. The contents of the iterator are unwound on demand into a cache array to allow re-generating the cells when you scroll up. It’s a little flickery but I’m not sure if that’s a browser issue or if it could be resolved by increasing the amount of extra cells rendered outside the visible area.

**Comment by j-f1** [comment]
Is it worth updating this to fix the merge conflicts or should I close?

**Comment by mbostock** [comment]
I would probably close for now — I think this will be difficult to land and I apologize for not having more time to look at it.

**Pull request closed without merging**